### PR TITLE
[Seq] Add `seq.fifo` ODS and rationale

### PR DESCRIPTION
**Pull request by mortbopet**
The `seq.fifo` operation intends to capture the semantics of a FIFO which eventually map to some form of on-chip resources. By having a FIFO abstraction, we provide an abstraction that can be targeted for target-specialized implementations, as well as default behavioral lowerings (based on `seq.hlmem`).

## Option 1: A `seq.fifo` op

The `seq.fifo` in- and output values consists of:
- **Inputs**:
  - clock, reset
  - input data
  - read/write enable
- **Outputs**:
  - output data
  - full, almost full, empty, almost empty flags

To start with, the FIFO operation is configurable with the following parameters:
1. Depth (cycles)
2. Differing in- and output widths (e.g. write `i32`, read 2x`i16`).
3. Almost empty/full thresholds (optional). If not provided, these will
    be asserted when the FIFO is empty/full.

Generally speaking, additional attributes can be added freely in order to further specialize the configuration (that is, provide information to a specific lowering pass) so that should take care of any lowering-specific information that needs to be provided.  
However, this op should ideally capture all possible in/outputs ports that we want (.. do we want dual clock support ?).

Like `seq.hlmem` there are no guarantees that all possible fifo configuration are able to be lowered. Available lowering passes will pattern match on the requested fifo configuration - it is then up to the user to ensure that they're requesting a configuration which is realizable by some pass.

## Option 2: A `seq.fifo` and `seq.flag_fifo` op

The `seq.flag_fifo` would have the FIFO flags built in. Is it more appropriate (from an IR/compiler point of view) to have a single `count` output, and it is then the users responsibility to implement flags (`cmpi` ops) based on this count? We could also have two operations, a `seq.flag_fifo` and `seq.fifo` wherein the former will lower to the latter.

## Option 3: Fifo ports attached to an `seq.hlmem`

In this model, the `seq.hlmem` continues to be a memory declaration. However, we'll add additional fifo read/write ports - different ports may have different properties or in/output values (just like our existing read and write ops).

```mlir
%handle = seq.hlmem %clk, %rst : !seq.hlmem<4xi32>
%out, %full, %almost_full = seq.fifo.flagged_read %handle, %rden
%empty, %almost_empty = seq.fifo.flagged_write %handle, %wren
```

Additionally, this model would support mixing fifo access with normal read/write access to a given backing memory - something which may be needed in certain edge cases.

Let me know your thoughts - I hope this issue can be a place where we can land on a well-designed IR representation of a FIFO... and then actual lowering can come down the road.

### CODE REVIEW

File: docs/Dialects/Seq/RationaleSeq.md [hunk 1]
```diff
@@ -299,3 +299,28 @@ specialization is needed) attached to the memory symbol.
...
+
+## The FIFO operation
+The `seq.fifo` operation intends to capture the semantics of a FIFO which
+eventually map to some form of on-chip resources. By having a FIFO abstraction,
+we provide an abstraction that can be targeted for target-specialized implementations,
+as well as default behavioral lowerings (based on `seq.hlmem`).
+
+The FIFO interface consists of:
+- **Inputs**:
+  - clock, reset
```
Comment (darthscsi) [inline docs/Dialects/Seq/RationaleSeq.md:311]: What exactly reset does has will have to be specified.

File: docs/Dialects/Seq/RationaleSeq.md [hunk 2]
```diff
@@ -299,3 +299,28 @@ specialization is needed) attached to the memory symbol.
...
+  - input data
+  - read/write enable
+- **Outputs**:
+  - output data
+  - full, almost full, empty, almost empty flags
```
Comment (darthscsi) [inline docs/Dialects/Seq/RationaleSeq.md:316]: all of these or some based on parameters?

**Review by darthscsi — COMMENTED** [review]
Generally looks ok.  I don't know if all the flags are necessary all the time?  My reading is that depending on parameters, the almost flags may be tied to the absolute flags.

**Comment by darthscsi** [comment]
As a first pass, a fifo with the flag outputs and parameterized by the almost full threshold seems like a good start.  I like that better than having a count output as it admits more implementations.  I don't know if differing read and write sizes are appropriate for a first pass.

I don't have any objections to the basic theory of adding something in these lines as long as there is an immediate use case for it.

**Comment by teqdruid** [comment]
Any interest in reviving this? Could be helpful in the immediate future.

I think Option 1 (like you have here) is the better fit for the intended use.

**Comment by mortbopet** [comment]
> Any interest in reviving this? Could be helpful in the immediate future.

... 👀 yes... once my backlog of DC and handshake stuff also starts to clear up...

**Comment by mortbopet** [comment]
Reviving this:
1. removing different in/output sizes - input size == output size
2. almost full/almost empty signals are now optional, and present based on the existence of the associated attributes in the op.

File: docs/Dialects/Seq/RationaleSeq.md (hunk 1 above)
Comment (mortbopet) [reply docs/Dialects/Seq/RationaleSeq.md:310]: Not sure i understand - could you elaborate?

File: include/circt/Dialect/Seq/SeqOps.td [hunk 3]
```diff
@@ -155,6 +155,55 @@ def FirRegOp : SeqOp<"firreg",
...
+
+  let arguments = (ins
+    HWIntegerType:$input, I1:$rdEn, I1:$wrEn, I1:$clk, I1:$rst,
+    ConfinedAttr<I64Attr, [IntMinValue<1>]>:$depth,
+    OptionalAttr<ConfinedAttr<I64Attr, [IntMinValue<0>]>>:$almostFullThreshold,
+    OptionalAttr<ConfinedAttr<I64Attr, [IntMinValue<0>]>>:$almostEmptyThreshold
+  );
+
+  let results = (outs
+    HWIntegerType:$output, I1:$full, I1:$empty, Optional<I1>:$almostFull, Optional<I1>:$almostEmpty);
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/SeqOps.td:188]: nit: line length?

File: include/circt/Dialect/Seq/SeqOps.td [hunk 4]
```diff
@@ -155,6 +155,55 @@ def FirRegOp : SeqOp<"firreg",
...
+    AttrSizedResultSegments
+  ]> {
+  let summary = "A high-level FIFO operation";
+  let description = [{
+    This operation represents a high-level abstraction of a FIFO. Access to the
+    FIFO is structural, and thus may be composed with other core RTL dialect
+    operations.
+    The fifo operation is configurable with the following parameters:
+    1. Depth (cycles)
+    2. Differing in- and output widths
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/SeqOps.td:171]: Not true anymore.

File: include/circt/Dialect/Seq/SeqOps.td [hunk 5]
```diff
@@ -155,6 +155,55 @@ def FirRegOp : SeqOp<"firreg",
...
+    OptionalAttr<ConfinedAttr<I64Attr, [IntMinValue<0>]>>:$almostFullThreshold,
+    OptionalAttr<ConfinedAttr<I64Attr, [IntMinValue<0>]>>:$almostEmptyThreshold
+  );
+
+  let results = (outs
+    HWIntegerType:$output, I1:$full, I1:$empty, Optional<I1>:$almostFull, Optional<I1>:$almostEmpty);
+
+  let builders = [
+    OpBuilder<(ins "Value":$clk, "Value":$rst, "StringRef":$symName,
+                   "llvm::ArrayRef<int64_t>":$shape, "Type":$elementType)>
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/SeqOps.td:192]: nit: `$shape` -> `$depth` ... ?

File: include/circt/Dialect/Seq/SeqOps.td [hunk 6]
```diff
@@ -155,6 +155,55 @@ def FirRegOp : SeqOp<"firreg",
...
+    ConfinedAttr<I64Attr, [IntMinValue<1>]>:$depth,
+    OptionalAttr<ConfinedAttr<I64Attr, [IntMinValue<0>]>>:$almostFullThreshold,
+    OptionalAttr<ConfinedAttr<I64Attr, [IntMinValue<0>]>>:$almostEmptyThreshold
+  );
+
+  let results = (outs
+    HWIntegerType:$output, I1:$full, I1:$empty, Optional<I1>:$almostFull, Optional<I1>:$almostEmpty);
+
+  let builders = [
+    OpBuilder<(ins "Value":$clk, "Value":$rst, "StringRef":$symName,
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/SeqOps.td:191]: Does this op have a symbol? I don't see it anywhere and I don't see a need.

File: include/circt/Dialect/Seq/SeqOps.td [hunk 7]
```diff
@@ -155,6 +155,55 @@ def FirRegOp : SeqOp<"firreg",
...
+
+  let assemblyFormat = [{
+    `depth` $depth
+    custom<FIFOAFThreshold>($almostFullThreshold, type($almostFull))
+    custom<FIFOAEThreshold>($almostEmptyThreshold, type($almostEmpty))
+    `in` $input `rdEn` $rdEn `wrEn` $wrEn `clk` $clk `rst` $rst attr-dict `:` type($input)
+  }];
+  let hasVerifier = 1;
+  let skipDefaultBuilders = 1;
+
```
Comment (teqdruid) [inline include/circt/Dialect/Seq/SeqOps.td:203]: uber-nit: ditch the empty newlines

File: lib/Dialect/Seq/SeqOps.cpp [hunk 8]
```diff
@@ -219,6 +219,81 @@ void HLMemOp::build(OpBuilder &builder, OperationState &result, Value clk,
...
+//===----------------------------------------------------------------------===//
+// FIFOOp
+//===----------------------------------------------------------------------===//
+
+// Flag threshold custom directive
+static ParseResult parseFIFOFlagThreshold(OpAsmParser &parser,
+                                          IntegerAttr &threshold,
+                                          Type &outputFlagType,
+                                          StringRef directive) {
+  // look for an optional "almost_full $threshold>" group.
```
Comment (teqdruid) [inline lib/Dialect/Seq/SeqOps.cpp:231]: Comment accuracy

File: lib/Dialect/Seq/SeqOps.cpp [hunk 9]
```diff
@@ -219,6 +219,81 @@ void HLMemOp::build(OpBuilder &builder, OperationState &result, Value clk,
...
+                                "almost_full");
+}
+
+ParseResult parseFIFOAEThreshold(OpAsmParser &parser, IntegerAttr &threshold,
+                                 Type &outputFlagType) {
+  return parseFIFOFlagThreshold(parser, threshold, outputFlagType,
+                                "almost_empty");
+}
+
+static void printFIFOFlagThreshold(OpAsmPrinter &p, Operation *op,
```
Comment (teqdruid) [inline lib/Dialect/Seq/SeqOps.cpp:258]: As a rule, if the amount of extra code is more than the code you're sharing, don't share it.

File: lib/Dialect/Seq/SeqOps.cpp [hunk 10]
```diff
@@ -589,11 +664,11 @@ OpFoldResult FirRegOp::fold(FoldAdaptor adaptor) {
 
   // If the register is held in permanent reset, replace it with its reset
   // value. This works trivially if the reset is asynchronous and therefore
-  // level-sensitive, in which case it will always immediately assume the reset
```
Comment (teqdruid) [inline lib/Dialect/Seq/SeqOps.cpp:592]: Unrelated formatting

**Review by teqdruid — APPROVED** [review]
lgtm!
